### PR TITLE
[SDF-28][SDF-30] Shader program

### PR DESCRIPTION
**Pull request by kubazuch**

### CODE REVIEW

**Comment by kubazuch** [comment]
There's itsy-bitsy bug regarding CRLF vs LF new lines present. It requires refactoring file loading, which I'M NOT in the mood of doing rn 🐱

**Comment by kubazuch** [comment]
Ooo la-la, looking spiffy 👀
image

**Comment by migoox** [comment]
> There's an itsy-bitsy bug present, regarding CRLF vs LF new lines. It requires refactoring of file loading, which I'M NOT in the mood of doing rn 🐱  

I will fix that👍

File: .clang-tidy [hunk 1]
```diff
@@ -19,10 +18,10 @@ Checks: >
   -cppcoreguidelines-pro-bounds-pointer-arithmetic,
   -cppcoreguidelines-pro-type-reinterpret-cast,
   -cppcoreguidelines-non-private-member-variables-in-classes,
+  -cppcoreguidelines-pro-type-union-access
```
Comment (migoox) [inline .clang-tidy:21]: Add `,`

File: libresin/libresin/core/shader.cpp [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+  auto stop     = clock::now();
+  auto duration = duration_cast<std::chrono::milliseconds>(stop - start);
+  Logger::debug("Shader {} recompilation took {}", shader_name_, duration);
+}
+
+void ShaderProgram::link_program() {
+  glLinkProgram(program_id_);
+
+  GLint status = 0;
+  glGetProgramiv(program_id_, GL_LINK_STATUS, &status);
```
Comment (migoox) [inline libresin/libresin/core/shader.cpp:39]: could be a static function that gets program id, `GL_LINK_STATUS`/`GL_VALIDATE_STATUS` and returns std::optional string or something, same for `glGetShaderiv`

File: libresin/libresin/core/shader.cpp [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+
+  glDetachShader(program_id_, vertex_shader);
+  glDetachShader(program_id_, fragment_shader);
+
+  glDeleteShader(vertex_shader);
+  glDeleteShader(fragment_shader);
+}
+
+ComputeShaderProgram::ComputeShaderProgram(std::string_view name, ShaderResource compute_shader)
+    : ShaderProgram(name), compute_shader_(std::move(compute_shader)) {
```
Comment (migoox) [inline libresin/libresin/core/shader.cpp:139]: Validate if the shader resource contains a compute shader.

File: libresin/libresin/core/shader.cpp [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+    std::string info(static_cast<GLuint>(length), '\0');
+    glGetShaderInfoLog(shader, length, &length, info.data());
+
+    log_throw(ShaderCreationException(get_shader_type_name(type), shader_name_, std::move(info)));
+  }
+
+  return shader;
+}
+
+RenderingShaderProgram::RenderingShaderProgram(std::string_view name, ShaderResource vertex_resource,
```
Comment (migoox) [inline libresin/libresin/core/shader.cpp:109]: Validate if the shader resource contains a vertex/fragment shader

File: libresin/tests/files_helper.hpp [hunk 5]
```diff
@@ -24,8 +26,11 @@ ::testing::AssertionResult IsFileContentEqual(const char* expected_expr, const c
   }
 
   std::string file_content((std::istreambuf_iterator<char>(file)), std::istreambuf_iterator<char>());
+  file_content = std::regex_replace(file_content, std::regex("\r\n"), "\n");
 
   if (file_content != content) {
+    std::println("|{}|", file_content);
```
Comment (migoox) [inline libresin/tests/files_helper.hpp:32]: Use `::testing::AssertionFailure() << `

File: libresin/tests/files_helper.hpp [hunk 6]
```diff
@@ -24,8 +26,11 @@ ::testing::AssertionResult IsFileContentEqual(const char* expected_expr, const c
   }
 
   std::string file_content((std::istreambuf_iterator<char>(file)), std::istreambuf_iterator<char>());
+  file_content = std::regex_replace(file_content, std::regex("\r\n"), "\n");
```
Comment (migoox) [inline libresin/tests/files_helper.hpp:29]: It should be handled now in the `libresin`

File: resin/resin/event/window_events.hpp [hunk 7]
```diff
@@ -10,11 +10,20 @@ class WindowCloseEvent : public Event<EventType::WindowCloseEvent> {
  public:
   EVENT_NAME(WindowCloseEvent)
 
-  std::string to_string() const override { return name(); }
+  std::string to_string() const override { return std::string(name()); }
```
Comment (migoox) [inline resin/resin/event/window_events.hpp:13]: `name()` returns `std::string_view`, so the `to_string` could also return the `std::string_view`, no need for the additional heap allocation, the caller may use the `std::string(...)` if they want to become an owner

File: libresin/tests/files_helper.hpp (hunk 5 above)
Comment (kubazuch) [reply libresin/tests/files_helper.hpp:32]: this was purely for debug and idk if shouldn't be deleted entirely?

File: resin/resin/event/window_events.hpp (hunk 7 above)
Comment (kubazuch) [reply resin/resin/event/window_events.hpp:13]: I vividly remember changing this to string_view. I must've dreamt it lol

File: libresin/tests/files_helper.hpp (hunk 5 above)
Comment (migoox) [reply libresin/tests/files_helper.hpp:32]: if so then delete it

File: resin/resin/event/window_events.hpp (hunk 7 above)
Comment (kubazuch) [reply resin/resin/event/window_events.hpp:13]: actually not viable due to 
```c++
std::string WindowResizeEvent::to_string() const override {
  std::stringstream ss;
  ss << name() << ": " << width_ << " x " << height_;
  return ss.str();
}
```
which has to remain a string

**Comment by kubazuch** [comment]
This is a major step forward for this project I fear. 

I've seen this week people are taking this project and really holding space with that and feeling power in that. 
I've seen it on a couple posts, I don't know how widespread, but you know -- I am in queer media, so that's my.. you know...

**Review by migoox — APPROVED** [review]

**Review by adamieck — APPROVED** [review]

**Pull request merged**